source/agora/consensus/validation/GenesisBlock.d

```diff
@@ -0,0 +1,216 @@
+
+    Copyright:
+        Copyright (c) 2019-2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.validation.GenesisBlock;
+
+import agora.common.Amount;
+import agora.common.crypto.Key;
+import agora.common.Hash;
+import agora.common.Types;
+import agora.common.Set;
+import agora.consensus.data.Block;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Transaction;
+import VEn = agora.consensus.validation.Enrollment;
+import VTx = agora.consensus.validation.Transaction;
+import agora.consensus.data.UTXOSet;
+import agora.consensus.Genesis;
+
+import std.algorithm;
+
+version (unittest)
+{
+    import agora.common.crypto.ECC;
+    import agora.common.crypto.Key;
+    import agora.common.crypto.Schnorr;
+    import agora.common.Hash;
+    import agora.consensus.data.Enrollment;
+    import agora.consensus.Genesis;
+    import agora.consensus.data.Transaction;
+    import agora.utils.Test;
+}
+
+/*******************************************************************************
+
+    Check the validity of a genesisBlock.
+
+    A genesisBlock is considered valid if:
+        - Block height is 0
+        - The previous block hash of the block must be empty
+        - The number of transactions in the block is above zero and less than
+          Block.TxsInBlock
+        - Transactions are ordered by their hash value
+        - All the transactions pass validation, which implies:
+            - Transaction type in block should only be Payment or Freeze
+            - There is only one input for the transaction
+            - The signature value of the input of the transaction is genuine
+            - There is more than one output of the transaction
+            - The output of the transaction should not be empty at address
+            - The output.value of the transaction must be in accordance
+              with the `Amount` rule
+        - The merkle root in the header matches the re-built merkle tree root
+        - All the enrollments pass validation, which implies:
+            - The Enrolments should be arranged in ascending order by utxo_key
+            - Signature are authentic
+            - The enrollments refer to freeze tx's in this block
+            - The freeze tx's spend an output which must be found with
+              the findUTXO() delegate
+            - Freeze transactions referenced in registration should not be
+              double-spend
+
+    Params:
+        block = The genesisBlock to check
+
+    Returns:
+        `null` if the genesisBlock is valid, a string explaining the reason it
+        is invalid otherwise.
+
+*******************************************************************************/
+
+public string isInvalidReason (const ref Block genesisBlock) nothrow @safe
```

Comment:
This should not be a public function, and it should not be an overload.

Reply:
Apply the suggestion.